File: frontend/src/map/geography/projection.js
/** Versioned equirectangular world projection matching the Python reference. */
import { createGeographicCoordinate } from "./contracts.js";

export const PROJECTION_ID = "projection:novegeo:equirectangular-world";
export const PROJECTION_VERSION = 1;
export const PROJECTION_TOLERANCE = 1e-8;

export function projectCoordinate(longitude, latitude) {
  const coordinate = createGeographicCoordinate(longitude, latitude);
  return Object.freeze({
    x: (coordinate.longitude + 180) / 360,
    y: (90 - coordinate.latitude) / 180,
    projectionId: PROJECTION_ID,
    projectionVersion: PROJECTION_VERSION,
  });
}

export function unprojectCoordinate(projected) {
  if (!projected || projected.projectionId !== PROJECTION_ID || projected.projectionVersion !== PROJECTION_VERSION) {
    throw new Error("projected coordinate uses an incompatible projection contract");
  }
  const x = Number(projected.x);
  const y = Number(projected.y);
  if (!Number.isFinite(x) || !Number.isFinite(y) || x < 0 || x > 1 || y < 0 || y > 1) {
    throw new RangeError("normalized projected coordinates must be between zero and one");
  }
  return createGeographicCoordinate(x * 360 - 180, 90 - y * 180);
}

File: infrastructure/geography/projection.py
"""Versioned equirectangular world projection shared with the browser."""
from __future__ import annotations

from dataclasses import dataclass

from .contracts import GeographicCoordinate, ProjectedCoordinate


@dataclass(frozen=True, slots=True)
class EquirectangularWorldProjection:
    projection_id: str = "projection:novegeo:equirectangular-world"
    version: int = 1
    tolerance: float = 1e-8

    def forward(self, coordinate: GeographicCoordinate) -> ProjectedCoordinate:
        return ProjectedCoordinate(
            x=(coordinate.longitude + 180.0) / 360.0,
            y=(90.0 - coordinate.latitude) / 180.0,
            projection_id=self.projection_id,
            projection_version=self.version,
        )

    def inverse(self, coordinate: ProjectedCoordinate) -> GeographicCoordinate:
        if coordinate.projection_id != self.projection_id or coordinate.projection_version != self.version:
            raise ValueError("projected coordinate uses an incompatible projection contract")
        if not 0.0 <= coordinate.x <= 1.0 or not 0.0 <= coordinate.y <= 1.0:
            raise ValueError("normalized projected coordinates must be between zero and one")
        return GeographicCoordinate(
            longitude=coordinate.x * 360.0 - 180.0,
            latitude=90.0 - coordinate.y * 180.0,
        )
